oak/server/rust/oak_runtime/src/node/grpc/client.rs

```diff
@@ -0,0 +1,258 @@
+            invocation.send_response(response)?;
+        }
+    }
+
+    // Creates a TLS connection to an external gRPC service.
+    async fn connect(&mut self) -> Result<GrpcRequestHandler, OakStatus> {
+        debug!("Connecting to {} [{}]", self.domain_name, self.address);
+
+        // Create a TLS configuration.
+        let tls_config = ClientTlsConfig::new()
+            .domain_name(self.domain_name.to_string())
+            .ca_certificate(self.root_tls_certificate.clone());
+
+        // Connect to a remote gRPC service.
+        let connection = Channel::builder(self.address.clone())
+            .tls_config(tls_config)
+            .connect()
+            .await
+            .map_err(|error| {
+                warn!(
+                    "Couldn't connect to {} [{}]: {:?}",
+                    self.domain_name, self.address, error
+                );
+                OakStatus::ErrInternal
+            })?;
+
+        debug!("Connected to {} [{}]", self.domain_name, self.address);
+        Ok(GrpcRequestHandler::new(connection))
+    }
+
+    /// Read a gRPC invocation from the [`GrpcClientNode::receiver`].
+    fn read_invocation(&self) -> Result<Invocation, OakStatus> {
+        let message = self.receiver.receive()?;
+        Invocation::create(&self.runtime, message)
+    }
+}
+
+/// Oak Node implementation for the gRPC client pseudo-Node.
+impl Node for GrpcClientNode {
+    fn start(self: Box<Self>) -> Result<JoinHandle<()>, OakStatus> {
+        let thread_handle = thread::Builder::new()
+            .name(self.to_string())
+            .spawn(move || self.worker_thread())
+            .expect("Failed to spawn thread");
+        Ok(thread_handle)
+    }
+}
+
+/// Sends gRPC requests to an external gRPC service and returns gRPC responses.
+struct GrpcRequestHandler {
+    /// A gRPC client dispatcher that wraps a gRPC connection and encodes/decodes messages via a
+    /// provided codec.
+    dispatcher: tonic::client::Grpc<tonic::transport::channel::Channel>,
+}
+
+impl GrpcRequestHandler {
+    fn new(connection: tonic::transport::channel::Channel) -> Self {
+        Self {
+            dispatcher: tonic::client::Grpc::new(connection),
+        }
+    }
+
+    /// Checks whether the [`GrpcRequestHandler::dispatcher`] is able to accept new requests.
+    async fn ready(&mut self) -> Result<(), OakStatus> {
+        self.dispatcher.ready().await.map_err(|error| {
+            error!("Service was not ready: {}", error);
+            OakStatus::ErrInternal
+        })?;
+        Ok(())
+    }
+
+    /// Sends an unary gRPC request wrapped in [`GrpcResponse`] to an external gRPC service.
+    ///
+    /// [`GrpcResponse`]: oak_abi::proto::oak::encap::GrpcResponse
+    async fn unary_request(&mut self, request: GrpcRequest) -> Result<GrpcResponse, OakStatus> {
+        self.ready().await?;
+        let codec = VecCodec::default();
+        let path = http::uri::PathAndQuery::from_maybe_shared(bytes::Bytes::from(
```

Comment:
It might be clearer to use something like `parse` (e.g. https://docs.rs/http/0.2.1/http/uri/struct.PathAndQuery.html#examples-3) or `from_str` here.